### DISCUSSION TITLE
When using interactive mode, Why do fabric create upper canvas for event handling instead of binding events to the lower canvas?

### DISCUSSION THREAD

**Discussion by tianmuji** [fabricjs/fabric.js#9689] [General]
I am trying to migrate fabricjs to HarmonyOS 

`  
protected createUpperCanvas() {
    const { el: lowerCanvasEl } = this.lower;
    const el = createCanvasElement();
    // we assign the same classname of the lowerCanvas
    el.className = lowerCanvasEl.className;
    // but then we remove the lower-canvas specific className
    el.classList.remove('lower-canvas');
    // we add the specific upper-canvas class
    el.classList.add('upper-canvas');
    el.setAttribute('data-fabric', 'top');
    el.style.cssText = lowerCanvasEl.style.cssText;
    el.setAttribute('draggable', 'true');
    return el;
  }
`

I see this code in CanvasDOMManager.ts，can I delete the upper canvas and bind event listeners to the lower canvas?

Can anyone help?
Thanks in advance

**Comment by ShaMan123** [comment]
It is used for interactive stuff such as
Brushes
Text selection

If you do not use there methods then it might be possible to remove it

**Reply by tianmuji** [reply]
Thank you for you help!
I just keep to two canvas.

when I try to migrate brush features, I got a problem
Where do fabric init freeDrawingBrush? 

In the project, I can not find it.
but if I download bundled js file, I see 
    _initInteractive: function() {
      this._currentTransform = null;
      this._groupSelector = null;
      this._initWrapperElement();
      this._createUpperCanvas();
      this._initEventListeners();

      this._initRetinaScaling();

      this.freeDrawingBrush = fabric.PencilBrush && new fabric.PencilBrush(this);

      this.calcOffset();
    },

Could you please help me find the place where freeDrawingBrush initialized?

**Reply by tianmuji** [reply]
https://github.com/fabricjs/fabric.js/blob/69724653d012cc4d677bf59913058286a52bad30/src/canvas/SelectableCanvas.ts#L294

it is defined in selectableCanvas.ts

**Reply by tianmuji** [reply]
I solved it
I should init a brush and pass the reference to freeDrawingBrush
